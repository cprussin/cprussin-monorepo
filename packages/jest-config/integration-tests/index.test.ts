import { exec } from "node:child_process";
import { rm, readFile } from "node:fs/promises";
import path from "node:path";
import { promisify } from "node:util";

import type { FormattedTestResults } from "@jest/test-result";

const ONE_SECOND_IN_MS = 1000;

const TEST_DIR = path.join(__dirname, "__fixtures__", "test-package");
const NODE_MODULES = path.join(TEST_DIR, "node_modules");
const TEST_RESULTS = path.join(TEST_DIR, "test-results.json");

const execAsync = promisify(exec);

type TestResult = FormattedTestResults["testResults"][number];
type AssertionResult = TestResult["assertionResults"][number];

// For some reason, `testFilePath` shows up in the test results but doesn't
// exist in the types that Jest distributes, so let's patch it in so typescript
// doesn't fail the code below.
type PatchedFormattedTestResults = Omit<FormattedTestResults, "testResults"> & {
  testResults: (Omit<TestResult, "assertionResults"> & {
    assertionResults: (AssertionResult & {
      testFilePath?: string;
    })[];
  })[];
};

describe("integration", () => {
  beforeEach(async () => {
    await rm(NODE_MODULES, { recursive: true, force: true });
    await rm(TEST_RESULTS, { force: true });
  });

  afterEach(async () => {
    await rm(NODE_MODULES, { recursive: true });
    await rm(TEST_RESULTS);
  });

  it(
    "installs & works",
    async () => {
      await execAsync("pnpm --ignore-workspace i", { cwd: TEST_DIR });
      await execAsync(
        "pnpm exec jest --json --outputFile ./test-results.json .",
        {
          cwd: TEST_DIR,
          // eslint-disable-next-line n/no-process-env
          env: { ...process.env, FORCE_COLOR: "0" },
        },
      );
      const testResults = JSON.parse(
        await readFile(TEST_RESULTS, "utf8"),
      ) as PatchedFormattedTestResults;
      expect(normalizeResults(testResults)).toMatchSnapshot();
    },
    120 * ONE_SECOND_IN_MS,
  );
});

const normalizeResults = (
  results: PatchedFormattedTestResults,
): PatchedFormattedTestResults => ({
  ...results,
  startTime: 0,
  testResults: results.testResults
    .sort((a, b) => a.name.localeCompare(b.name))
    .map((testResult) => ({
      ...testResult,
      startTime: 0,
      endTime: 0,
      name: testResult.name.replace(TEST_DIR, "mocked-root-dir"),
      assertionResults: testResult.assertionResults.map((assertionResult) => ({
        ...assertionResult,
        duration: 0,
        ...("testFilePath" in assertionResult && {
          testFilePath: assertionResult.testFilePath.replace(
            TEST_DIR,
            "mocked_root_dir",
          ),
        }),
      })),
    })),
});
