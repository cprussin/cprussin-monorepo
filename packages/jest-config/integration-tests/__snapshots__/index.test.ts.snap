// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration installs & works 1`] = `
{
  "numFailedTestSuites": 0,
  "numFailedTests": 0,
  "numPassedTestSuites": 8,
  "numPassedTests": 8,
  "numPendingTestSuites": 4,
  "numPendingTests": 1,
  "numRuntimeErrorTestSuites": 0,
  "numTodoTests": 0,
  "numTotalTestSuites": 12,
  "numTotalTests": 9,
  "openHandles": [],
  "snapshot": {
    "added": 0,
    "didUpdate": false,
    "failure": false,
    "filesAdded": 0,
    "filesRemoved": 0,
    "filesRemovedList": [],
    "filesUnmatched": 0,
    "filesUpdated": 0,
    "matched": 0,
    "total": 0,
    "unchecked": 0,
    "uncheckedKeysByFile": [],
    "unmatched": 0,
    "updated": 0,
  },
  "startTime": 0,
  "success": true,
  "testResults": [
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "numPassingAsserts": 1,
          "status": "passed",
          "testFilePath": "mocked_root_dir/eslint.config.js",
          "title": "ESLint",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/eslint.config.js",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "",
          "numPassingAsserts": 1,
          "status": "passed",
          "title": "",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/eslint.config.js",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "numPassingAsserts": 1,
          "status": "passed",
          "testFilePath": "mocked_root_dir/jest.config.js",
          "title": "ESLint",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/jest.config.js",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "",
          "numPassingAsserts": 1,
          "status": "passed",
          "title": "",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/jest.config.js",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "numPassingAsserts": 0,
          "status": "skipped",
          "testFilePath": "mocked_root_dir/package.json",
          "title": "ESLint",
        },
      ],
      "coverage": {},
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/package.json",
      "startTime": 0,
      "status": "skipped",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "",
          "numPassingAsserts": 1,
          "status": "passed",
          "title": "",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/package.json",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "prettier",
          "numPassingAsserts": 0,
          "status": "skipped",
          "title": "prettier",
        },
      ],
      "coverage": {},
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/pnpm-lock.yaml",
      "startTime": 0,
      "status": "skipped",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "numPassingAsserts": 0,
          "status": "skipped",
          "testFilePath": "mocked_root_dir/src/index.test.ts",
          "title": "ESLint",
        },
      ],
      "coverage": {},
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/src/index.test.ts",
      "startTime": 0,
      "status": "skipped",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [
            "sayHello",
          ],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "sayHello says hello",
          "invocations": 1,
          "location": null,
          "numPassingAsserts": 1,
          "retryReasons": [],
          "status": "passed",
          "title": "says hello",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/src/index.test.ts",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "",
          "numPassingAsserts": 1,
          "status": "passed",
          "title": "",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/src/index.test.ts",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "numPassingAsserts": 0,
          "status": "skipped",
          "testFilePath": "mocked_root_dir/src/index.ts",
          "title": "ESLint",
        },
      ],
      "coverage": {},
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/src/index.ts",
      "startTime": 0,
      "status": "skipped",
      "summary": "",
    },
    {
      "assertionResults": [
        {
          "ancestorTitles": [],
          "duration": 0,
          "failureDetails": [],
          "failureMessages": [],
          "fullName": "",
          "numPassingAsserts": 1,
          "status": "passed",
          "title": "",
        },
      ],
      "endTime": 0,
      "message": "",
      "name": "mocked-root-dir/src/index.ts",
      "startTime": 0,
      "status": "passed",
      "summary": "",
    },
  ],
  "wasInterrupted": false,
}
`;
